File: leetcode/104-maximum-depth-of-binary-tree/main.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// dfs, iterative

type Stack struct {
	Node  *TreeNode
	Depth int
}

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	maxDepth := 0
	var stack []*Stack
	stack = append(stack, &Stack{root, 1})
	for len(stack) > 0 {
		pop := stack[len(stack)-1]
		stack = stack[0 : len(stack)-1]
		if pop.Depth > maxDepth {
			maxDepth = pop.Depth
		}
		if pop.Node.Left != nil {
			stack = append(stack, &Stack{pop.Node.Left, pop.Depth + 1})
		}
		if pop.Node.Right != nil {
			stack = append(stack, &Stack{pop.Node.Right, pop.Depth + 1})
		}
	}
	return maxDepth
}

// dfs, recursive

func maxDepth1(root *TreeNode) int {
	if root == nil {
		return 0
	}
	left := maxDepth1(root.Left)
	right := maxDepth1(root.Right)
	if left > right {
		return left + 1
	}
	return right + 1
}

// bfs, iterative

type Queue struct {
	Node  *TreeNode
	Depth int
}

func maxDepth2(root *TreeNode) int {
	if root == nil {
		return 0
	}
	maxDepth := 0
	var queue []*Queue
	queue = append(queue, &Queue{root, 1})
	for len(queue) > 0 {
		n := len(queue)
		for i := 0; i < n; i++ {
			head := queue[0]
			queue = queue[1:len(queue)]
			if head.Depth > maxDepth {
				maxDepth = head.Depth
			}
			if head.Node.Left != nil {
				queue = append(queue, &Queue{head.Node.Left, head.Depth + 1})
			}
			if head.Node.Right != nil {
				queue = append(queue, &Queue{head.Node.Right, head.Depth + 1})
			}
		}
	}
	return maxDepth
}

func main() {
	// 		5
	//	2		8
	// 1 3 7 9
	//				10
	root := &TreeNode{5,
		&TreeNode{2,
			&TreeNode{1, nil, nil},
			&TreeNode{3, nil, nil},
		},
		&TreeNode{8,
			&TreeNode{7, nil, nil},
			&TreeNode{9,
				nil,
				&TreeNode{10, nil, nil},
			},
		},
	}
	ans := maxDepth(root)
	fmt.Println(ans)
	ans = maxDepth1(root)
	fmt.Println(ans)
	ans = maxDepth2(root)
	fmt.Println(ans)
}

File: leetcode/104-maximum-depth-of-binary-tree/main.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
    1st approach: recursive dfs

    Time    O(n)
    Space   O(h)
    24 ms, faster than 97.03%
"""


class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return self.dfs(root)

    def dfs(self, node):
        if node == None:
            return 0
        left = self.dfs(node.left)
        right = self.dfs(node.right)
        return max(left, right)+1


"""
    2nd approach: iterative dfs
    
    Time    O(n)
    Space   O(h)
"""


class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root == None:
            return 0
        res = 0
        stack = [(root, 1)]
        while len(stack) > 0:
            node, steps = stack.pop(0)
            res = max(res, steps)
            if node.left != None:
                stack.append((node.left, steps + 1))
            if node.right != None:
                stack.append((node.right, steps + 1))
        return res


"""
    iterative bfs
    
    Time    O(n)
    Space   O(h)
"""


class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root == None:
            return 0
        res = 0
        q = [(root, 1)]
        while len(q) > 0:
            pop, depth = q.pop(0)
            res = max(res, depth)
            if pop.left != None:
                q.append((pop.left, depth + 1))
            if pop.right != None:
                q.append((pop.right, depth + 1))
        return res
